### PR TITLE
KAZOO-3669: As a user I want the forgot password to not change my password until I click on a link in an email so that it works like other systems I am used to

### PR DESCRIPTION
**Pull request by fenollp**
WIP

### CODE REVIEW

**Comment by lazedo** [comment]
@fenollp is this ready ? want to squash these commits ?

**Comment by fenollp** [comment]
@lazedo Needs work and waiting on https://2600hz.atlassian.net/browse/UI-1819

File: applications/crossbar/src/modules/cb_user_auth.erl [hunk 1]
```diff
@@ -16,17 +17,20 @@
...
 -define(ACCT_MD5_LIST, <<"users/creds_by_md5">>).
 -define(ACCT_SHA1_LIST, <<"users/creds_by_sha">>).
--define(USERNAME_LIST, <<"users/list_by_username">>).
+-define(LIST_BY_RESETID, <<"users/list_by_resetid">>).
 -define(DEFAULT_LANGUAGE, <<"en-us">>).
 -define(USER_AUTH_TOKENS, whapps_config:get_integer(?CONFIG_CAT, <<"user_auth_tokens">>, 35)).
 
 -define(RECOVERY, <<"recovery">>).
+-define(RESET_ID, <<"resetid">>).
+-define(PWD_RESETID_SIZE, 40).
```
Comment (jamesaimonetti) [inline applications/crossbar/src/modules/cb_user_auth.erl]: May as well put this in a `whapps_config:get_integer/3` call

File: applications/crossbar/src/modules/cb_user_auth.erl [hunk 2]
```diff
@@ -106,6 +112,7 @@ authenticate_nouns(_Nouns) -> 'false'.
 -spec validate(cb_context:context()) -> cb_context:context().
 -spec validate(cb_context:context(), path_token()) -> cb_context:context().
 validate(Context) ->
+    lager:debug(">>> validate/1"),
```
Comment (jamesaimonetti) [inline applications/crossbar/src/modules/cb_user_auth.erl]: Still needed?

File: applications/crossbar/src/modules/cb_user_auth.erl [hunk 3]
```diff
@@ -114,20 +121,38 @@ validate(Context) ->
     end.
 
 validate(Context, ?RECOVERY) ->
-    cb_context:validate_request_data(<<"user_auth_recovery">>, Context, fun maybe_recover_user_password/1);
+    Verb = cb_context:req_verb(Context),
+    validate_recovery(Context, Verb);
 validate(Context, Token) ->
+    lager:debug(">>> validate/2 "),
```
Comment (jamesaimonetti) [inline applications/crossbar/src/modules/cb_user_auth.erl]: Still needed?

File: applications/crossbar/src/modules/cb_user_auth.erl [hunk 4]
```diff
@@ -114,20 +121,38 @@ validate(Context) ->
...
     Context1 = cb_context:set_account_db(Context, ?KZ_TOKEN_DB),
     maybe_get_auth_token(Context1, Token).
 
+-spec validate_recovery(cb_context:context(), http_method()) -> cb_context:context().
+validate_recovery(Context, ?HTTP_PUT) ->
+    cb_context:validate_request_data(<<"user_auth_recovery">>, Context, fun maybe_recover_user_password/1);
+validate_recovery(Context, ?HTTP_POST) ->
+    lager:debug(">>> validate/2 RECOVERY"),
```
Comment (jamesaimonetti) [inline applications/crossbar/src/modules/cb_user_auth.erl]: Either of these log lines needed?

File: applications/crossbar/src/modules/cb_user_auth.erl [hunk 5]
```diff
@@ -353,45 +397,65 @@ maybe_load_username(Account, Context) ->
...
-    Password = wh_util:rand_hex_binary(16),
-    {MD5, SHA1} = cb_modules_util:pass_hashes(wh_json:get_value(<<"username">>, JObj), Password),
-    Email = wh_json:get_value(<<"email">>, JObj),
-
-    JObj1 = wh_json:set_values([{<<"pvt_md5_auth">>, MD5}
-                                ,{<<"pvt_sha1_auth">>, SHA1}
-                                ,{<<"require_password_update">>, 'true'}
-                               ], JObj),
+-spec reset_users_password__step_1(cb_context:context()) -> cb_context:context().
+reset_users_password__step_1(Context) ->
```
Comment (jamesaimonetti) [inline applications/crossbar/src/modules/cb_user_auth.erl]: Two `__`?

**Comment by jamesaimonetti** [comment]
What process is going to clean up the expired/unused reset links? Look to the `crossbar_cleanup` module perhaps? See how `cb_vmboxes` uses it.

**Pull request closed without merging**